components/tidb_query/src/codec/convert.rs

```diff
@@ -2077,4 +2075,564 @@ mod tests {
+                1,
+                false,
+                Ok(Decimal::zero()),
+                None,
+                false,
+                false,
+                false,
+            ),
+        ];
+
+        for (
+            input,
+            origin_flen,
+            origin_decimal,
+            res_flen,
+            res_decimal,
+            is_unsigned,
+            expect,
+            warning_err_code,
+            in_dml,
+            overflow_as_warning,
+            truncate_as_warning,
+        ) in cs
+        {
+            // check origin_flen and origin_decimal
+            let (f, d) = input.prec_and_frac();
+            let log = format!(
+                    "input: {}, origin_flen: {}, origin_decimal: {}, actual flen: {}, actual decimal: {}",
+                    input, origin_flen, origin_decimal, f, d
+                );
+            assert_eq!(f, origin_flen, "{}", log);
+            assert_eq!(d, origin_decimal, "{}", log);
+
+            // run test case
+            let ctx_in_dml_flag = vec![Flag::IN_INSERT_STMT, Flag::IN_UPDATE_OR_DELETE_STMT];
+            for in_dml_flag in ctx_in_dml_flag {
+                // make ctx
+                let mut flag: Flag = Flag::default();
+                if overflow_as_warning {
+                    flag |= Flag::OVERFLOW_AS_WARNING;
+                }
+                if truncate_as_warning {
+                    flag |= Flag::TRUNCATE_AS_WARNING;
+                }
+                if in_dml {
+                    flag |= in_dml_flag;
+                }
+                let cfg = Arc::new(EvalConfig::from_flag(flag));
+                let mut ctx = EvalContext::new(cfg);
+
+                // make field_type
+                let mut rft = FieldType::default();
+                let fta = rft.as_mut_accessor();
+                fta.set_flen(res_flen);
+                fta.set_decimal(res_decimal);
+                if is_unsigned {
+                    fta.set_flag(FieldTypeFlag::UNSIGNED);
+                }
+
+                // call produce_dec_with_specified_tp
+                let r = produce_dec_with_specified_tp(&mut ctx, input.clone(), &rft);
+
+                // make log
+                let rs = r.as_ref().map(|x| x.to_string());
+                let expect_str = expect.as_ref().map(|x| x.to_string());
+                let log =
+                    format!(
+                            "input: {}, origin_flen: {}, origin_decimal: {}, \
+                     res_flen: {}, res_decimal: {}, is_unsigned: {}, \
+                     in_dml: {}, in_dml_flag(if in_dml is false, it will take no effect): {:?}, \
+                     expect: {:?}, expect: {:?}",
+                            input, origin_flen, origin_decimal, res_flen, res_decimal,
+                            is_unsigned, in_dml, in_dml_flag, expect_str, rs
+                        );
+
+                // check result
+                match expect {
+                    Ok(ref d) => {
```

Comment:
I think it's better to avoid `ref` in pattern match if possible? 

```rust 
assert_eq!(&r.unwrap(), &d, "{}", log);
```

And the nested match below can be condensed e.g: 
```rust
Err(Error::Eval(_, _)) => { ... }
_ => unreachable!(log)
```